go-controller/pkg/ovn/controller/network_qos/network_qos_controller.go

```diff
@@ -0,0 +1,452 @@
+	}
+
+	klog.V(5).Info("Starting Node Network QoS workers")
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNQOSNodeWorker(wg)
+			}, time.Second, stopCh)
+		}()
+	}
+
+	<-stopCh
+
+	klog.Infof("Shutting down controller %s", c.controllerName)
+	c.nqosQueue.ShutDown()
+	c.nqosNamespaceQueue.ShutDown()
+	c.nqosPodQueue.ShutDown()
+	c.teardownMetricsCollector()
+	wg.Wait()
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same object
+// at the same time.
+func (c *Controller) runNQOSWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSNamespaceWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSNamespaceWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSPodWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSPodWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSNodeWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSNodeWorkItem(wg) {
+	}
+}
+
+// handlers
+
+// onNQOSAdd queues the NQOS for processing.
+func (c *Controller) onNQOSAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Adding Network QoS %s", key)
+	c.nqosQueue.Add(key)
+}
+
+// onNQOSUpdate updates the NQOS Selector in the cache and queues the NQOS for processing.
+func (c *Controller) onNQOSUpdate(oldObj, newObj interface{}) {
+	oldNQOS := oldObj.(*networkqosapi.NetworkQoS)
+	newNQOS := newObj.(*networkqosapi.NetworkQoS)
+
+	// don't process resync or objects that are marked for deletion
+	if oldNQOS.ResourceVersion == newNQOS.ResourceVersion ||
+		!newNQOS.GetDeletionTimestamp().IsZero() {
+		return
+	}
+	if reflect.DeepEqual(oldNQOS.Spec, newNQOS.Spec) {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err == nil {
+		// updates to NQOS object should be very rare, once put in place they usually stay the same
+		klog.V(4).Infof("Updating Network QoS %s: nqosSpec %v",
+			key, newNQOS.Spec)
+		c.nqosQueue.Add(key)
+	}
+}
+
+// onNQOSDelete queues the NQOS for processing.
+func (c *Controller) onNQOSDelete(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Deleting Network QoS %s", key)
+	c.nqosQueue.Add(key)
+}
+
+// onNQOSNamespaceAdd queues the namespace for processing.
+func (c *Controller) onNQOSNamespaceAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(5).Infof("Adding Namespace in Network QoS controller %s", key)
```

Comment:
why is this one 5, but the nqos ones are 4

Reply:
removed.